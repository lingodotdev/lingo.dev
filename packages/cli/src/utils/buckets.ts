import _ from 'lodash';
import path from 'path';
import * as glob from 'glob';
import { ReplexicaCLIError } from './errors';
import { I18nConfig } from '@replexica/spec';
import { bucketTypeSchema } from '@replexica/spec';
import Z from 'zod';

export function getBuckets(i18nConfig: I18nConfig) {
  let result = Object
    .entries(i18nConfig.buckets)
    .map(([bucketType, bucketEntry]) => ({
      type: bucketType as Z.infer<typeof bucketTypeSchema>,
      pathPatterns: extractPathPatterns(i18nConfig.locale.source, bucketEntry.include, bucketEntry?.exclude),
    }));

  return result;
}

function extractPathPatterns(
  sourceLocale: string,
  include: string[],
  exclude?: string[],
) {
  const includedPatterns = include.flatMap((pattern) => expandPlaceholderedGlob(pattern, sourceLocale));
  const excludedPatterns = exclude?.flatMap((pattern) => expandPlaceholderedGlob(pattern, sourceLocale));
  const result = _.difference(includedPatterns, excludedPatterns ?? []);
  return result;
}


// Path expansion
function expandPlaceholderedGlob(_pathPattern: string, sourceLocale: string): string[] {
  // Throw if pathPattern is an absolute path
  const absolutePathPattern = path.resolve(_pathPattern);
  const pathPattern = path.relative(process.cwd(), absolutePathPattern);
  // Throw if pathPattern points outside the current working directory
  if (path.relative(process.cwd(), pathPattern).startsWith('..')) {
    throw new ReplexicaCLIError({
      message: `Invalid path pattern: ${pathPattern}. Path pattern must be within the current working directory.`,
      docUrl: "invalidPathPattern"
    });
  }
  // Throw error if pathPattern contains "**" â€“ we don't support recursive path patterns
  if (pathPattern.includes('**')) {
    throw new ReplexicaCLIError({
      message: `Invalid path pattern: ${pathPattern}. Recursive path patterns are not supported.`,
      docUrl: 'invalidPathPattern'
    });
  }
  // Throw error if pathPattern contains "[locale]" several times
  if (pathPattern.split('[locale]').length > 2) {
    throw new ReplexicaCLIError({
      message: `Invalid path pattern: ${pathPattern}. Path pattern must contain at most one "[locale]" placeholder.`,
      docUrl: "invalidPathPattern"
    });
  }
  // Break down path pattern into parts
  const pathPatternChunks = pathPattern.split(path.sep);
  // Find the index of the segment containing "[locale]"
  const localeSegmentIndex = pathPatternChunks.findIndex((segment) => segment.includes('[locale]'));
  // Find the position of the "[locale]" placeholder within the segment
  const localePlaceholderIndex = pathPatternChunks[localeSegmentIndex]?.indexOf('[locale]') ?? -1;
  // substitute [locale] in pathPattern with sourceLocale
  const sourcePathPattern = pathPattern.replace(/\[locale\]/g, sourceLocale);
  // get all files that match the sourcePathPattern
  const sourcePaths = glob
    .sync(sourcePathPattern, { follow: true, withFileTypes: true })
    .filter((file) => file.isFile() || file.isSymbolicLink())
    .map((file) => file.fullpath())
    .map((fullpath) => path.relative(process.cwd(), fullpath));
  // transform each source file path back to [locale] placeholder paths
  const placeholderedPaths = sourcePaths.map((sourcePath) => {
    const sourcePathChunks = sourcePath.split(path.sep);
    if (localeSegmentIndex >= 0 && localePlaceholderIndex >= 0) {
      const placeholderedPathChunk = sourcePathChunks[localeSegmentIndex];
      const placeholderedSegment =
        placeholderedPathChunk.substring(0, localePlaceholderIndex)
        + '[locale]'
        + placeholderedPathChunk.substring(localePlaceholderIndex + sourceLocale.length)
        ;
      sourcePathChunks[localeSegmentIndex] = placeholderedSegment;
    }
    const placeholderedPath = sourcePathChunks.join(path.sep);
    return placeholderedPath;
  });
  // return the placeholdered paths
  return placeholderedPaths;
}
