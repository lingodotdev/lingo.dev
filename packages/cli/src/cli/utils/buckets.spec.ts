import { describe, it, expect, vi, beforeEach } from "vitest";
import { getBuckets, parsePatternTemplate } from "./buckets";
import { glob, Path } from "glob";

vi.mock("glob", () => ({
  glob: {
    sync: vi.fn(),
  },
}));

describe("parsePatternTemplate", () => {
  it("captures globstars and placeholder segments", () => {
    expect(parsePatternTemplate("src/**/[locale]/**/messages.json")).toEqual([
      {
        kind: "segment",
        original: "src",
        parts: [{ kind: "literal", value: "src" }],
        hasPlaceholder: false,
        hasGlob: false,
      },
      { kind: "globstar", original: "**" },
      {
        kind: "segment",
        original: "[locale]",
        parts: [{ kind: "placeholder", name: "locale" }],
        hasPlaceholder: true,
        hasGlob: false,
      },
      { kind: "globstar", original: "**" },
      {
        kind: "segment",
        original: "messages.json",
        parts: [{ kind: "literal", value: "messages.json" }],
        hasPlaceholder: false,
        hasGlob: false,
      },
    ]);
  });

  it("identifies placeholders embedded within segments", () => {
    expect(parsePatternTemplate("src/messages.[locale].json")).toEqual([
      {
        kind: "segment",
        original: "src",
        parts: [{ kind: "literal", value: "src" }],
        hasPlaceholder: false,
        hasGlob: false,
      },
      {
        kind: "segment",
        original: "messages.[locale].json",
        parts: [
          { kind: "literal", value: "messages." },
          { kind: "placeholder", name: "locale" },
          { kind: "literal", value: ".json" },
        ],
        hasPlaceholder: true,
        hasGlob: false,
      },
    ]);
  });

  it("marks glob syntax distinct from placeholders", () => {
    expect(
      parsePatternTemplate("src/modules/@(core|marketing)-[locale].json"),
    ).toEqual([
      {
        kind: "segment",
        original: "src",
        parts: [{ kind: "literal", value: "src" }],
        hasPlaceholder: false,
        hasGlob: false,
      },
      {
        kind: "segment",
        original: "modules",
        parts: [{ kind: "literal", value: "modules" }],
        hasPlaceholder: false,
        hasGlob: false,
      },
      {
        kind: "segment",
        original: "@(core|marketing)-[locale].json",
        parts: [
          { kind: "glob", value: "@(core|marketing)-" },
          { kind: "placeholder", name: "locale" },
          { kind: "literal", value: ".json" },
        ],
        hasPlaceholder: true,
        hasGlob: true,
      },
    ]);
  });
});

describe("getBuckets", () => {
  beforeEach(() => {
    vi.mocked(glob.sync).mockReset();
  });

  const makeI18nConfig = (include: any[]) => ({
    $schema: "https://lingo.dev/schema/i18n.json",
    version: 0,
    locale: {
      source: "en",
      targets: ["fr", "es"],
    },
    buckets: {
      json: {
        include,
      },
    },
  });

  it("should return correct buckets", () => {
    mockGlobSync(["src/i18n/en.json"], ["src/translations/en/messages.json"]);

    const i18nConfig = makeI18nConfig([
      "src/i18n/[locale].json",
      "src/translations/[locale]/messages.json",
    ]);
    const buckets = getBuckets(i18nConfig);
    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/i18n/[locale].json", delimiter: null },
          {
            pathPattern: "src/translations/[locale]/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("should return correct buckets for paths with asterisk", () => {
    mockGlobSync(
      [
        "src/translations/landing.en.json",
        "src/translations/app.en.json",
        "src/translations/email.en.json",
      ],
      [
        "src/locale/landing/messages.en.json",
        "src/locale/app/data.en.json",
        "src/locale/email/custom.en.json",
      ],
      [
        "src/i18n/landing/en.messages.json",
        "src/i18n/app/en.data.json",
        "src/i18n/email/en.custom.json",
      ],
      [
        "src/i18n/data-landing-en-strings/en.messages.json",
        "src/i18n/data-app-en-strings/en.data.json",
        "src/i18n/data-email-en-strings/en.custom.json",
      ],
    );

    const i18nConfig = makeI18nConfig([
      "src/translations/*.[locale].json",
      "src/locale/*/*.[locale].json",
      "src/i18n/*/[locale].*.json",
      "src/i18n/data-*-[locale]-*/[locale].*.json",
    ]);
    const buckets = getBuckets(i18nConfig);
    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/translations/landing.[locale].json",
            delimiter: null,
          },
          {
            pathPattern: "src/translations/app.[locale].json",
            delimiter: null,
          },
          {
            pathPattern: "src/translations/email.[locale].json",
            delimiter: null,
          },
          {
            pathPattern: "src/locale/landing/messages.[locale].json",
            delimiter: null,
          },
          { pathPattern: "src/locale/app/data.[locale].json", delimiter: null },
          {
            pathPattern: "src/locale/email/custom.[locale].json",
            delimiter: null,
          },
          {
            pathPattern: "src/i18n/landing/[locale].messages.json",
            delimiter: null,
          },
          { pathPattern: "src/i18n/app/[locale].data.json", delimiter: null },
          {
            pathPattern: "src/i18n/email/[locale].custom.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/i18n/data-landing-[locale]-strings/[locale].messages.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/i18n/data-app-[locale]-strings/[locale].data.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/i18n/data-email-[locale]-strings/[locale].custom.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("should return correct bucket with delimiter", () => {
    mockGlobSync(["src/i18n/en.json"]);
    const i18nConfig = makeI18nConfig([
      { path: "src/i18n/[locale].json", delimiter: "-" },
    ]);
    const buckets = getBuckets(i18nConfig);
    expect(buckets).toEqual([
      {
        type: "json",
        paths: [{ pathPattern: "src/i18n/[locale].json", delimiter: "-" }],
      },
    ]);
  });

  it("should return bucket with multiple locale placeholders", () => {
    mockGlobSync(
      ["src/i18n/en/en.json"],
      ["src/en/translations/en/messages.json"],
    );
    const i18nConfig = makeI18nConfig([
      "src/i18n/[locale]/[locale].json",
      "src/[locale]/translations/[locale]/messages.json",
    ]);
    const buckets = getBuckets(i18nConfig);
    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/i18n/[locale]/[locale].json", delimiter: null },
          {
            pathPattern: "src/[locale]/translations/[locale]/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("restores locale placeholder when using recursive globstar patterns", () => {
    mockGlobSync([
      "src/modules/core/auth/en/strings/messages.json",
      "src/modules/marketing/en/strings/dashboard.json",
    ]);

    const i18nConfig = makeI18nConfig([
      "src/modules/**/[locale]/strings/*.json",
    ]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/modules/core/auth/[locale]/strings/messages.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/modules/marketing/[locale]/strings/dashboard.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("restores placeholder when extglob wraps the locale segment", () => {
    mockGlobSync(["src/modules/core-en.json", "src/modules/marketing-en.json"]);

    const i18nConfig = makeI18nConfig([
      "src/modules/@(core|marketing)-[locale].json",
    ]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/modules/core-[locale].json", delimiter: null },
          {
            pathPattern: "src/modules/marketing-[locale].json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("restores placeholder when brace expansion surrounds locale segment", () => {
    mockGlobSync([
      "src/modules/core/en/strings/messages.json",
      "src/modules/marketing/en/strings/dashboard.json",
    ]);

    const i18nConfig = makeI18nConfig([
      "src/modules/{core,marketing}/[locale]/strings/*.json",
    ]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/modules/core/[locale]/strings/messages.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/modules/marketing/[locale]/strings/dashboard.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("preserves glob character classes around locale placeholder", () => {
    mockGlobSync(["src/files/id-en.json"]);

    const i18nConfig = makeI18nConfig(["src/files/??-[locale].json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/files/id-[locale].json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("supports globstar at the beginning of the pattern", () => {
    mockGlobSync([
      "src/modules/core/en/messages.json",
      "src/modules/marketing/en/dashboard.json",
    ]);

    const i18nConfig = makeI18nConfig(["**/[locale]/*.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/modules/core/[locale]/messages.json",
            delimiter: null,
          },
          {
            pathPattern: "src/modules/marketing/[locale]/dashboard.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("supports multiple globstars surrounding the locale segment", () => {
    mockGlobSync([
      "src/modules/core/services/en/api/messages.json",
      "src/modules/marketing/en/email/templates/messages.json",
    ]);

    const i18nConfig = makeI18nConfig(["src/**/[locale]/**/messages.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/modules/core/services/[locale]/api/messages.json",
            delimiter: null,
          },
          {
            pathPattern:
              "src/modules/marketing/[locale]/email/templates/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("supports globstar segments after the locale placeholder", () => {
    mockGlobSync(["src/i18n/en/deep/messages.json"]);

    const i18nConfig = makeI18nConfig(["src/i18n/[locale]/**/messages.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/i18n/[locale]/deep/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("supports globstar leading directly into the locale file name", () => {
    mockGlobSync(["src/en.json", "src/translations/en.json"]);

    const i18nConfig = makeI18nConfig(["**/[locale].json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/[locale].json", delimiter: null },
          {
            pathPattern: "src/translations/[locale].json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("supports trailing globstar before the file extension", () => {
    mockGlobSync(["src/files/en/report.json", "src/files/en/app.json"]);

    const i18nConfig = makeI18nConfig(["src/files/[locale]/**.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/files/[locale]/report.json",
            delimiter: null,
          },
          {
            pathPattern: "src/files/[locale]/app.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("handles consecutive globstars before the locale segment", () => {
    mockGlobSync(["src/a/b/en/messages.json", "src/en/messages.json"]);

    const i18nConfig = makeI18nConfig(["src/**/**/[locale]/messages.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/a/b/[locale]/messages.json",
            delimiter: null,
          },
          {
            pathPattern: "src/[locale]/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("prioritizes the earliest matching locale segment when duplicates exist", () => {
    mockGlobSync(["src/en/module/en/messages.json"]);

    const i18nConfig = makeI18nConfig(["src/**/[locale]/**/messages.json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          {
            pathPattern: "src/[locale]/module/en/messages.json",
            delimiter: null,
          },
        ],
      },
    ]);
  });

  it("deduplicates overlapping include patterns", () => {
    mockGlobSync(["src/i18n/en.json"]);
    mockGlobSync(["src/i18n/en.json"]);

    const i18nConfig = makeI18nConfig([
      "src/i18n/**/[locale].json",
      "src/i18n/[locale].json",
    ]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [{ pathPattern: "src/i18n/[locale].json", delimiter: null }],
      },
    ]);
  });

  it("keeps distinct entries for matching paths with different delimiters", () => {
    mockGlobSync(["src/i18n/en.json"]);
    mockGlobSync(["src/i18n/en.json"]);

    const i18nConfig = makeI18nConfig([
      { path: "src/i18n/[locale].json", delimiter: "-" },
      { path: "src/i18n/[locale].json", delimiter: "_" },
    ]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/i18n/[locale].json", delimiter: "-" },
          { pathPattern: "src/i18n/[locale].json", delimiter: "_" },
        ],
      },
    ]);
  });

  it("excludes entries matching both path pattern and delimiter", () => {
    mockGlobSync(["src/i18n/en.json"]);
    mockGlobSync(["src/i18n/en.json"]);
    mockGlobSync(["src/i18n/en.json"]);

    const i18nConfig = makeI18nConfig([
      { path: "src/i18n/[locale].json", delimiter: "-" },
      { path: "src/i18n/[locale].json", delimiter: "_" },
    ]);
    i18nConfig.buckets.json.exclude = [
      { path: "src/i18n/[locale].json", delimiter: "-" },
    ];

    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [{ pathPattern: "src/i18n/[locale].json", delimiter: "_" }],
      },
    ]);
  });

  it("restores placeholder when locale appears multiple times in a segment", () => {
    mockGlobSync(["src/files/en-en.json"]);

    const i18nConfig = makeI18nConfig(["src/files/[locale]-[locale].json"]);
    const buckets = getBuckets(i18nConfig);

    expect(buckets).toEqual([
      {
        type: "json",
        paths: [
          { pathPattern: "src/files/[locale]-[locale].json", delimiter: null },
        ],
      },
    ]);
  });

  it("throws when pattern resolves outside of the current working directory", () => {
    const i18nConfig = makeI18nConfig(["../outside/[locale].json"]);

    expect(() => getBuckets(i18nConfig)).toThrowError(
      /Invalid path pattern: \.{2}\//,
    );
  });

  describe("placeholder restoration matrix", () => {
    const cases: Array<{
      name: string;
      include: any[];
      globResults: string[][];
    }> = [
      {
        name: "basic segment placeholder",
        include: ["src/[locale]/messages.json"],
        globResults: [["src/en/messages.json"]],
      },
      {
        name: "placeholder inside file extension",
        include: ["src/messages.[locale].json"],
        globResults: [["src/messages.en.json"]],
      },
      {
        name: "placeholder with prefix and suffix in same segment",
        include: ["src/files/pre[locale]post.json"],
        globResults: [["src/files/preenpost.json"]],
      },
      {
        name: "duplicated placeholder within a single segment",
        include: ["src/files/[locale]-[locale].json"],
        globResults: [["src/files/en-en.json"]],
      },
      {
        name: "multiple placeholder segments",
        include: ["src/[locale]/module/[locale]/messages.json"],
        globResults: [["src/en/module/en/messages.json"]],
      },
      {
        name: "static locale directory before placeholder",
        include: ["src/en/module/[locale]/messages.json"],
        globResults: [["src/en/module/en/messages.json"]],
      },
      {
        name: "static locale directory after placeholder",
        include: ["src/[locale]/module/en/messages.json"],
        globResults: [["src/en/module/en/messages.json"]],
      },
      {
        name: "globstar before placeholder with extra segments",
        include: ["src/**/[locale]/messages.json"],
        globResults: [["src/features/en/messages.json"]],
      },
      {
        name: "globstar before placeholder with zero extra segments",
        include: ["src/**/[locale]/messages.json"],
        globResults: [["src/en/messages.json"]],
      },
      {
        name: "globstar after placeholder",
        include: ["src/i18n/[locale]/**/messages.json"],
        globResults: [["src/i18n/en/deep/messages.json"]],
      },
      {
        name: "globstar surrounding placeholder with duplicate locale segment",
        include: ["src/**/[locale]/**/messages.json"],
        globResults: [["src/en/module/en/messages.json"]],
      },
      {
        name: "multiple placeholders separated by globstars",
        include: ["src/**/[locale]/**/[locale].json"],
        globResults: [["src/en/foo/en/en.json"]],
      },
      {
        name: "extglob wrapping locale placeholder",
        include: ["src/modules/@(core|marketing)-[locale].json"],
        globResults: [["src/modules/core-en.json"]],
      },
      {
        name: "brace expansion containing locale segment",
        include: ["src/modules/{core,marketing}/[locale]/strings/*.json"],
        globResults: [["src/modules/core/en/strings/messages.json"]],
      },
      {
        name: "character class adjacent to locale placeholder",
        include: ["src/files/??-[locale].json"],
        globResults: [["src/files/id-en.json"]],
      },
      {
        name: "placeholder between single-segment wildcards",
        include: ["src/*/[locale]/file.json"],
        globResults: [["src/app/en/file.json"]],
      },
      {
        name: "deep translations path with static locale duplication",
        include: ["src/**/[locale]/**/translations/[locale].json"],
        globResults: [["src/en/module/en/translations/en.json"]],
      },
    ];

    cases.forEach(({ name, include, globResults }) => {
      it(name, () => {
        mockGlobSync(...globResults);
        const i18nConfig = makeI18nConfig(include);
        const buckets = getBuckets(i18nConfig);

        expect(buckets).toMatchSnapshot();
      });
    });
  });
});

function mockGlobSync(...args: string[][]) {
  args.forEach((files) => {
    vi.mocked(glob.sync).mockReturnValueOnce(
      files.map(
        (file) => ({ isFile: () => true, fullpath: () => file }) as Path,
      ),
    );
  });
}
