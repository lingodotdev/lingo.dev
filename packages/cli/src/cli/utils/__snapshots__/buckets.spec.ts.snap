// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBuckets > placeholder restoration matrix > basic segment placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > brace expansion containing locale segment 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/modules/core/[locale]/strings/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > character class adjacent to locale placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/files/id-[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > deep translations path with static locale duplication 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/module/en/translations/[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > duplicated placeholder within a single segment 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/files/[locale]-[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > extglob wrapping locale placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/modules/core-[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > globstar after placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/i18n/[locale]/deep/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > globstar before placeholder with extra segments 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/features/[locale]/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > globstar before placeholder with zero extra segments 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > globstar surrounding placeholder with duplicate locale segment 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/module/en/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > multiple placeholder segments 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/module/[locale]/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > multiple placeholders separated by globstars 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/foo/en/[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > placeholder between single-segment wildcards 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/app/[locale]/file.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > placeholder inside file extension 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/messages.[locale].json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > placeholder with prefix and suffix in same segment 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/files/pre[locale]post.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > static locale directory after placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/[locale]/module/en/messages.json",
      },
    ],
    "type": "json",
  },
]
`;

exports[`getBuckets > placeholder restoration matrix > static locale directory before placeholder 1`] = `
[
  {
    "paths": [
      {
        "delimiter": null,
        "pathPattern": "src/en/module/[locale]/messages.json",
      },
    ],
    "type": "json",
  },
]
`;
